### PR TITLE
31 should retry on 5xx

### PR DESCRIPTION
**Pull request by jakubchrobak**
#### Summary
Provide retry on server errors.

#### Description
- I provided `getRetryOn5xxClient` function in `ClientConfigurationUtils`, that wraps `SphereClient` into a `RetrySphereClientDecorator`.
- I used that function in every sync's constructor while instantiating services.
- I provided up to 27 retries with durations between each retry `2s, 2s, 2s, 4s, 4s, 4s, 8s, ... , 512s`, for each failed request. 
- I wrote relevant unit tests.

#### Relevant Issues
https://github.com/commercetools/commercetools-sync-java/issues/31

#### Hints for Review
Please express opinion about retry strategy, and if it should be more customizable within a method parameters.

### CODE REVIEW

File: src/main/java/com/commercetools/sync/categories/CategorySync.java [hunk 1]
```diff
@@ -53,8 +54,8 @@
      */
     public CategorySync(@Nonnull final CategorySyncOptions syncOptions) {
         this(syncOptions,
-            new TypeServiceImpl(syncOptions.getCtpClient()),
-            new CategoryServiceImpl(syncOptions.getCtpClient()));
+            new TypeServiceImpl(getRetryOn5xxClient(syncOptions.getCtpClient())),
```
Comment (heshamMassoud) [inline src/main/java/com/commercetools/sync/categories/CategorySync.java]: I wouldn't do it here on every time we need to get the client, but rather one time when it's injected and bound to the options.

File: src/main/java/com/commercetools/sync/commons/utils/ClientConfigurationUtils.java [hunk 2]
```diff
@@ -1,22 +1,32 @@
...
 
+import static io.sphere.sdk.retry.RetryAction.ofScheduledRetry;
+import static io.sphere.sdk.retry.RetryPredicate.ofMatchingErrors;
+import static java.util.Collections.singletonList;
+
 public class ClientConfigurationUtils {
     private static BlockingSphereClient ctpClient;
     private static final long DEFAULT_TIMEOUT = 30;
     private static final TimeUnit DEFAULT_TIMEOUT_TIME_UNIT = TimeUnit.SECONDS;
+    private static final int RETRY_ON_5XX_ATTEMPTS_LIMIT = 27;
```
Comment (heshamMassoud) [inline src/main/java/com/commercetools/sync/commons/utils/ClientConfigurationUtils.java]: why 27?

Comment (heshamMassoud) [inline src/main/java/com/commercetools/sync/commons/utils/ClientConfigurationUtils.java]: Also why is it a global variable and not passed as a parameter to the method it's used in?

File: src/main/java/com/commercetools/sync/commons/utils/ClientConfigurationUtils.java [hunk 3]
```diff
@@ -46,6 +56,36 @@ public static BlockingSphereClient createClient(@Nonnull final SphereClientConfi
...
+    @Nonnull
+    public static SphereClient getRetryOn5xxClient(@Nonnull final SphereClient sphereClient) {
+        return RetrySphereClientDecorator.of(sphereClient, singletonList(RetryRule.of(
+            ofMatchingErrors(ServerErrorException.class),
+            ofScheduledRetry(RETRY_ON_5XX_ATTEMPTS_LIMIT, ClientConfigurationUtils::retryOn5xxDurationCalculator))));
+    }
+
+    /**
+     * Given a {@link RetryContext} it returns duration for a current attempt. The duration is calculated from a
+     * following equation: <pre>duration = 2 ^ (((retryAttemptNumber - 1) / 3) + 1) [seconds]</pre>
```
Comment (heshamMassoud) [inline src/main/java/com/commercetools/sync/commons/utils/ClientConfigurationUtils.java]: could you please explain the idea behind this strategy?

File: src/main/java/com/commercetools/sync/commons/utils/ClientConfigurationUtils.java (hunk 2 above)
Comment (jakubchrobak) [inline src/main/java/com/commercetools/sync/commons/utils/ClientConfigurationUtils.java]: 27 fits into proposed default strategy. More about it in a comment below.

File: src/main/java/com/commercetools/sync/commons/utils/ClientConfigurationUtils.java (hunk 3 above)
Comment (jakubchrobak) [inline src/main/java/com/commercetools/sync/commons/utils/ClientConfigurationUtils.java]: About the strategy: it is only my proposition. In fact I have no clue what retry strategy is wanted here. I have expressed my concerns some time ago in a issue https://github.com/commercetools/commercetools-sync-java/issues/31#issuecomment-306434605

This particular strategy seemed reasonable to me: first it performs few attempts with a little duration and then the time is increased. There is 27 attempts so it sum to almost 2 hours of waiting (e.g. when the server is down). 

Please share your opinion about what is expected here @heshamMassoud

File: src/main/java/com/commercetools/sync/categories/CategorySync.java (hunk 1 above)
Comment (jakubchrobak) [inline src/main/java/com/commercetools/sync/categories/CategorySync.java]: Ah, I forgot about the options. Good point :+1:

File: src/main/java/com/commercetools/sync/commons/utils/ClientConfigurationUtils.java (hunk 3 above)
Comment (heshamMassoud) [inline src/main/java/com/commercetools/sync/commons/utils/ClientConfigurationUtils.java]: retry strategy: https://github.com/sphereio/sphere-node-utils/blob/master/src/coffee/helpers/repeater.coffee#L90-L96

Provide through IT as an example, not function of sync..

**Pull request closed without merging**